# Популярные имена
**Цели:** <br>● провести предобработку данных, для последующего их удобного анализа
<br>
● проанализировать данные в соответсвии с имеющимися задачами
<br>
<br>
**Цели:** <br>● посчитать сколько детей в 2022 году получили такое же имя как у меня ("Анастасия").
<br>
● проверить истинность утверждения «Самыми популярными детскими именами на протяжении последних трех лет являются Александр и София (Софья)».
<br>
● вычислить на сколько процентов изменилась популярность женского имени Есения и мужского имени Сергей с начала 2022 года по сравнению с аналогичным периодом 2021 года?
<br>
<br>
**Входные данные:** <br>входные данные представляют собой файл csv, содержащий в себе информацию о различных спортсменах и их участии в олимпиадах.
<br>
<br>
**Используемые библиотеки:** <br>pandas
<br>
<br>
Для начала импортируем библиотеки

In [1]:
import pandas as pd
import numpy as np


Для предобработки и анализа мы имеем 2 файла. Один содержит статистику по женским именам, а другой по мужским. Разобъем наш проект на 2 глобальных части: в первой поработаем с данными по женским именам, а во второй со статистикой мужских имен.
<br>
<br>
## Статистика по женским именам
Считаем файл с данными, который мы предварительно сохранили. Заметим, что его кодировка отличается от привычной нам. ТАкже вместо запятых, характерных для файлов "csv", там используются ";". Учтем это при считывании. Запишем его в переменную df1:

In [2]:
df1 = pd.read_csv("/Users/anastasiakula/Desktop/инфа школьная/data-6269-2022-08-26.csv", encoding='Windows-1251', sep = ";")


Выведем основную информацию о данных:

In [3]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9101 entries, 0 to 9100
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  9101 non-null   object 
 1   Name                9101 non-null   object 
 2   NumberOfPersons     9101 non-null   object 
 3   global_id           9101 non-null   object 
 4   Year                9101 non-null   object 
 5   Month               9101 non-null   object 
 6   ID_en               1 non-null      object 
 7   Name_en             1 non-null      object 
 8   NumberOfPersons_en  1 non-null      object 
 9   Year_en             1 non-null      object 
 10  Month_en            1 non-null      object 
 11  Unnamed: 11         0 non-null      float64
dtypes: float64(1), object(11)
memory usage: 853.3+ KB


Описание данных:
● id - идентификатор<br>
● name - имя<br>
● NumberOfPersons - количество человек<br>
● global_id - глобальный идентификатор<br>
● Year - год<br>
● Month - месяц<br>
● ID_en - идентификатор (английский)<br>
● Name_en - имя (английский)<br>
● NumberOfPersons_en - количество человек (английский)<br>
● Year_en - год (английский)<br>
● Month_en - месяц (английский)<br>
● Unnamed: 11 - ?<br>
<br>
Можно заметить, что данные состоят из 9100 строчек и 12 столбиков. Также присутствуют пропуски в некоторых ячейках, есть также полностью пустые столбики. Заметим также, что типы данных у всех столбцов кроме "Unnamed: 11" - "object". Запомним это и вернемся к этому вопросу чуть позже.


Посмотрим какой вид имеют наши данные. Для этого выведем первые 10 строк:

In [4]:
df1.head(10)

ID        Name     NumberOfPersons  global_id  Year   Month  ID_en  \
0  Код         Имя  Количество человек  global_id   Год   Месяц  ID_en   
1    1       Мария                 252   37750254  2015  январь    NaN   
2    2   Анастасия                 224   37750255  2015  январь    NaN   
3    3        Анна                 190   37750256  2015  январь    NaN   
4    4     Варвара                 190   37750257  2015  январь    NaN   
5    5    Виктория                 186   37750258  2015  январь    NaN   
6  483      Амелия                   8   62367755  2015     май    NaN   
7  484      Камила                   8   62367756  2015     май    NaN   
8  485  Евангелина                   8   62367757  2015     май    NaN   
9  486     Альбина                   8   62367758  2015     май    NaN   

   Name_en  NumberOfPersons_en  Year_en  Month_en  Unnamed: 11  
0  Name_en  NumberOfPersons_en  Year_en  Month_en          NaN  
1      NaN                 NaN      NaN       NaN          NaN  
2      NaN                 NaN      NaN       NaN          NaN  
3      NaN                 NaN      NaN       NaN          NaN  
4      NaN                 NaN      NaN       NaN          NaN  
5      NaN                 NaN      NaN       NaN          NaN  
6      NaN                 NaN      NaN       NaN          NaN  
7      NaN                 NaN      NaN       NaN          NaN  
8      NaN                 NaN      NaN       NaN          NaN  
9      NaN                 NaN      NaN       NaN          NaN

Видно, что названия колонок записаны не в стиле **snake_case**. Исправим это:

In [5]:
df1.columns = df1.columns.str.lower()
df1.head(1)

id name     numberofpersons  global_id year  month  id_en  name_en  \
0  Код  Имя  Количество человек  global_id  Год  Месяц  ID_en  Name_en   

   numberofpersons_en  year_en  month_en  unnamed: 11  
0  NumberOfPersons_en  Year_en  Month_en          NaN

In [6]:
df1.columns 

Index(['id', 'name', 'numberofpersons', 'global_id', 'year', 'month', 'id_en',
       'name_en', 'numberofpersons_en', 'year_en', 'month_en', 'unnamed: 11'],
      dtype='object')

Теперь разберемся с лишними стообцами. Поймем, что нам не нужны следующие столбцы:<br>
    ● id<br>
    ● global_id<br>
    ● id_en<br>
    ● name_en<br>
    ● numberofpersons_en<br>
    ● year_en<br>
    ● month_en<br>
    ● unnamed: 11<br>
 <br>
Некоторые из них содержат только пустые ячейки, некоторые, такие как id, global_id не пригодяться для выполнения поставленных задач. Удалим их все:

In [7]:
df1 = df1.drop(["id"], axis=1)
df1 = df1.drop(["id_en"], axis=1)
df1 = df1.drop(["name_en"], axis=1)
df1 = df1.drop(["numberofpersons_en"], axis=1)
df1 = df1.drop(["year_en"], axis=1)
df1 = df1.drop(["global_id"], axis=1)
df1 = df1.drop(["month_en"], axis=1)
df1 = df1.drop(["unnamed: 11"], axis=1)
df1.head(5)

name     numberofpersons  year   month
0        Имя  Количество человек   Год   Месяц
1      Мария                 252  2015  январь
2  Анастасия                 224  2015  январь
3       Анна                 190  2015  январь
4    Варвара                 190  2015  январь

Поймем, что первая строчка для нас абсолютно бесполезна, т.к. содержит лишь перевод названия столбиков. Удалим ее:

In [8]:
df1 = df1.drop(index=[0])
df1.head(5)

name numberofpersons  year   month
1      Мария             252  2015  январь
2  Анастасия             224  2015  январь
3       Анна             190  2015  январь
4    Варвара             190  2015  январь
5   Виктория             186  2015  январь

Проверим наличие пропусков в каждом из столбиков:

In [9]:
df1.isna().mean()

name               0.0
numberofpersons    0.0
year               0.0
month              0.0
dtype: float64

Пропусков нет.

Проверим неявные дубликаты в колонке "year", для этого выведем все уникальные значения в этом столбце:

In [10]:
df1['year'].value_counts()


2015    1200
2016    1200
2017    1200
2018    1200
2019    1200
2020    1200
2021    1200
2022     700
Name: year, dtype: int64

Неявных дубликатов нет. Проверим колонку "month":

In [11]:
df1['month'].value_counts()

май         500
июнь        500
июль        500
август      500
январь      400
Апрель      400
Март        400
Февраль     400
Январь      400
Декабрь     400
Ноябрь      400
Октябрь     400
апрель      400
март        400
февраль     400
сентябрь    400
декабрь     300
ноябрь      300
октябрь     300
Май         300
Июнь        300
Июль        300
Сентябрь    300
Август      200
Name: month, dtype: int64

Неявные дубликаты есть. Заменим все значения на названия месяцев на русском с маленькой буквы:

In [12]:
df1.loc[:, 'month'] = df1.loc[:, 'month'].str.lower()
df1['month'].value_counts()

январь      800
май         800
июнь        800
июль        800
февраль     800
март        800
апрель      800
август      700
сентябрь    700
октябрь     700
ноябрь      700
декабрь     700
Name: month, dtype: int64

Проверить на наличие неявных дубликатов колонку "name" будет сложно, ведь существует много разных форм одного имени. Также проверять колонку "name" на уникальные значения бессмысленно, т.к. их очень много. Но для тех, информацию о которых я буду анализировать, я попробую это учесть. 
<br>
<br>
Проверим на наличие явных дубликатов:




In [13]:
df1.duplicated().sum()

0

Явных дубликатов нет. Перед тем как приступить к выполнению первой задачи, надо не забыть поменять типы данных для столбцов "year", "numberofpersons"

In [14]:
df1["year"] = df1["year"].astype(int)
df1["numberofpersons"] = df1["numberofpersons"].astype(int)

## Сколько людей в 2022 году получили имя "Анастасия"
Создадим df1_new, в который запишем данные, соответствующие фильтру (год = 2022, имя = Анастасия)

In [15]:
a = 2022
b = "Анастасия"
df1.head(10)
df1_new = df1.query('name == @b and year == @a')

df1_new.tail(20)


name  numberofpersons  year    month
8411  Анастасия               96  2022   январь
8511  Анастасия               94  2022  февраль
8617  Анастасия               80  2022     март
8716  Анастасия               71  2022   апрель
8812  Анастасия               86  2022      май
8916  Анастасия               80  2022     июнь
9013  Анастасия               92  2022     июль

In [16]:
ans = df1_new["numberofpersons"].sum()
print(ans)

599


Проверим, не использовалась ли сокращенная форма имени "Настя":

In [17]:
df1.query('name	 == "Настя"').name.count()

0

Эта самая популярная сокращенная форма имени не использовалась. Значит можно сделать вывод, что имя "Анастасия" получили в 2022 году 519 человек.
<br>
<br>
Я не смогу учесть все варианты формулировок имени "Анастасия" не просматривая данные вручную. Но когда я просмотрела их, других форм данного имени там не было.
<br>
<br>
## Проверить истинность утверждения «Самым популярным женским детским именем на протяжении последних трех лет является София (Софья)».
Создадим df1_new2, в который запишем данные соответсвующие критерию "год лежит в промежутке 2020-2022"`:

In [18]:
a = 2022
b = 2021
c = 2020
df1_new2 = df1.query('year == @a or year == @b or year == @c')
df1_new2.head(10)

name  numberofpersons  year   month
6001       Мария              180  2020  январь
6002       София              174  2020  январь
6003        Анна              162  2020  январь
6004      Полина              160  2020  январь
6005       Алиса              151  2020  январь
6006    Виктория              130  2020  январь
6007   Анастасия              127  2020  январь
6008  Александра              111  2020  январь
6009       Дарья              108  2020  январь
6010         Ева              105  2020  январь

Создадим df1_new3, где каждая строчка будет иметь название каждого имени, встречающегося в df1_new2, а в столбике "numberofpersons" будет сумма всех значений по этому имени.

In [19]:
df1_new3 = pd.pivot_table(df1_new2,
               values=["numberofpersons"],
               index =["name"],
               aggfunc=[np.sum])
df1_new3.head(-20)

sum
         numberofpersons
name                    
Аврора               534
Агата               1077
Аглая                120
Агния                216
Аделина              855
...                  ...
София               5795
Софья               1554
Стефания            1077
Сумая                193
Таисия              1352

[143 rows x 1 columns]

Отсортируем df1_new3. Тогда на последнем месте будет стоять строка с тем именем, которое использовалось больше всего раз за этот период.

In [20]:
df1_new4 = df1_new3.sort_values([('sum', 'numberofpersons')])
df1_new4.tail(10)


sum
           numberofpersons
name                      
Варвара               3364
Александра            3490
Елизавета             3502
Ева                   3925
Полина                3959
Виктория              4178
Алиса                 4471
Анна                  5307
Мария                 5519
София                 5795

Хоть мы и так видим, что София - наиболее популярное имя, посчитаем также сколько детей были названы "Софья"

In [21]:
df1_new3.loc["Софья"]

sum  numberofpersons    1554
Name: Софья, dtype: int64

Попробуем посмотреть не за весь период трех лет, а за каждый год отдельно. Код будем испольщовать такой же, тольк для удобства запишем его в функцию и вызовем ее 3 раза, применяя фильтр с каждым из годов.

In [22]:
a = 2022
b = 2021
c = 2020
def count_name(a):
    df1_new2 = df1.query('year == @a')
    df1_new3 = pd.pivot_table(df1_new2,
                   values=["numberofpersons"],
                   index =["name"],
                   aggfunc=[np.sum])
    df1_new4 = df1_new3.sort_values([('sum', 'numberofpersons')])
    c = df1_new3.loc["Софья"]
    print(c[-1])
    q = df1_new4.tail(2)
    return q

    
print(count_name(a))
print(count_name(b))
print(count_name(c))

303
                  sum
      numberofpersons
name                 
Мария            1129
София            1271
617
                  sum
      numberofpersons
name                 
Мария            2172
София            2322
634
                  sum
      numberofpersons
name                 
София            2202
Мария            2218


Проанализировав данные можно увидеть, что действительно в каждый из последних трех годов имя "София/Софья" лидировало по количесвту(в 2020 году имя "София" встречается 2202 раз, "Софья" 634 раза, в сумме 2836, что больше, чем количество людей с именем "Мария").

## Вычислить на сколько процентов изменилась популярность женского имени Есения  с начала 2022 года по сравнению с аналогичным периодом 2021 года? 
Для начала осознаем, что вообще от нас хотят. Мы берем промежуток с января по сентябрь 2022 года, считаем количество имени "Есения" в этот период. Потом берем промежуток с января по сентябрь 2021 года и опять считаем количество имени "Есения" в этот период. А далее говорим, что данные за 2021 год - 100%, посчитаем соклько процентов сейчас.
<br>
<br>
Для реализации также будем использовать фильтр, на этот раз выставив условие на год, месяц и имя. Посчитаем за 2022 год:

In [23]:
a = 2022
df1_new2 = df1.query('year == @a and name == "Есения"')
df1_new2.head(10)

name  numberofpersons  year    month
8426  Есения               44  2022   январь
8524  Есения               48  2022  февраль
8623  Есения               50  2022     март
8721  Есения               55  2022   апрель
8827  Есения               46  2022      май
8926  Есения               50  2022     июнь
9022  Есения               68  2022     июль

А тепеь просто просуммируем столбик "numberofpersons"

In [24]:
ans = df1_new2["numberofpersons"].sum()
print(ans)

361


Итого 361 ребенок с именем "Есения" за это период. Перейдем к 2021 году, рассматривая период с января по июль:

In [25]:
a = 2021
df1_new2 = df1.query('year == @a and name == "Есения" and month != "август" and month != "сентябрь" and month != "октябрь" and month != "ноябрь" and month != "декабрь"')
df1_new2.head(10)

name  numberofpersons  year    month
7232  Есения               32  2021   январь
7345  Есения               20  2021  февраль
7427  Есения               42  2021     март
7523  Есения               55  2021   апрель
7624  Есения               53  2021      май
7723  Есения               52  2021     июнь
7825  Есения               55  2021     июль

Просуммируем столбик "numberofpersons"

In [26]:
ans = df1_new2["numberofpersons"].sum()
print(ans)

309


Вычислим на какой процент увеличилось количество детей с именем "Есения", если сравнивать период 2021 года с января по июль и период 2022 года (те же месяцы):
309 = 100%
361 = x%

x = 361 * 100 / 309

In [27]:
print(361 * 100 / 309)

116.8284789644013


Таким образом процент возрос на 16.3.
С анализом женских имен мы закончили, переходим к данным по мудским именам.
## Статистика по мужским именам

In [28]:
df2 = pd.read_csv("/Users/anastasiakula/Desktop/инфа школьная/data-6271-2022-08-26.csv", encoding='Windows-1251', sep = ";")

In [29]:
df2.head()

ID       Name     NumberOfPersons  global_id  Year   Month  ID_en  \
0  Код        Имя  Количество человек  global_id   Год   Месяц  ID_en   
1    1  Александр                 253   37750243  2015  январь    NaN   
2    2     Михаил                 247   37750244  2015  январь    NaN   
3    3       Иван                 226   37750245  2015  январь    NaN   
4    4     Максим                 208   37750246  2015  январь    NaN   

   Name_en  NumberOfPersons_en  Year_en  Month_en  Unnamed: 11  
0  Name_en  NumberOfPersons_en  Year_en  Month_en          NaN  
1      NaN                 NaN      NaN       NaN          NaN  
2      NaN                 NaN      NaN       NaN          NaN  
3      NaN                 NaN      NaN       NaN          NaN  
4      NaN                 NaN      NaN       NaN          NaN

Данные выглядят точно также, как и в файле с женскими именами. Также присутствуют лишние столбики и строчка. Поэтому действия будут аналогичны или почти аналогичны тем, что выполнялись ранее на стадии предобработки данных по женским именам

Выведем основную информацию о данных:

In [30]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9101 entries, 0 to 9100
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  9101 non-null   object 
 1   Name                9101 non-null   object 
 2   NumberOfPersons     9101 non-null   object 
 3   global_id           9101 non-null   object 
 4   Year                9101 non-null   object 
 5   Month               9101 non-null   object 
 6   ID_en               1 non-null      object 
 7   Name_en             1 non-null      object 
 8   NumberOfPersons_en  1 non-null      object 
 9   Year_en             1 non-null      object 
 10  Month_en            1 non-null      object 
 11  Unnamed: 11         0 non-null      float64
dtypes: float64(1), object(11)
memory usage: 853.3+ KB


Ничем не отличаеся от первой базы данных

Выведем первые 10 строк:

In [31]:
df2.head(10)

ID       Name     NumberOfPersons  global_id  Year   Month  ID_en  \
0  Код        Имя  Количество человек  global_id   Год   Месяц  ID_en   
1    1  Александр                 253   37750243  2015  январь    NaN   
2    2     Михаил                 247   37750244  2015  январь    NaN   
3    3       Иван                 226   37750245  2015  январь    NaN   
4    4     Максим                 208   37750246  2015  январь    NaN   
5    5      Артём                 190   37750247  2015  январь    NaN   
6    6     Даниил                 171   62361364  2015  январь    NaN   
7    7    Дмитрий                 164   62361365  2015  январь    NaN   
8    8     Кирилл                 151   62361366  2015  январь    NaN   
9    9     Андрей                 130   62361367  2015  январь    NaN   

   Name_en  NumberOfPersons_en  Year_en  Month_en  Unnamed: 11  
0  Name_en  NumberOfPersons_en  Year_en  Month_en          NaN  
1      NaN                 NaN      NaN       NaN          NaN  
2      NaN                 NaN      NaN       NaN          NaN  
3      NaN                 NaN      NaN       NaN          NaN  
4      NaN                 NaN      NaN       NaN          NaN  
5      NaN                 NaN      NaN       NaN          NaN  
6      NaN                 NaN      NaN       NaN          NaN  
7      NaN                 NaN      NaN       NaN          NaN  
8      NaN                 NaN      NaN       NaN          NaN  
9      NaN                 NaN      NaN       NaN          NaN

Также, как и с первыми данными, названия колонок записаны не в стиле **snake_case**. Исправим это:

In [32]:
df2.columns = df2.columns.str.lower()
df2.head(1)

id name     numberofpersons  global_id year  month  id_en  name_en  \
0  Код  Имя  Количество человек  global_id  Год  Месяц  ID_en  Name_en   

   numberofpersons_en  year_en  month_en  unnamed: 11  
0  NumberOfPersons_en  Year_en  Month_en          NaN

In [33]:
df2.columns 

Index(['id', 'name', 'numberofpersons', 'global_id', 'year', 'month', 'id_en',
       'name_en', 'numberofpersons_en', 'year_en', 'month_en', 'unnamed: 11'],
      dtype='object')

Теперь разберемся с лишними стообцами. Они аналогичны лишним столбцам в первых данных. Удалим:

In [34]:
df2 = df2.drop(["id"], axis=1)
df2 = df2.drop(["id_en"], axis=1)
df2 = df2.drop(["name_en"], axis=1)
df2 = df2.drop(["numberofpersons_en"], axis=1)
df2 = df2.drop(["year_en"], axis=1)
df2 = df2.drop(["global_id"], axis=1)
df2 = df2.drop(["month_en"], axis=1)
df2 = df2.drop(["unnamed: 11"], axis=1)
df2.head(5)

name     numberofpersons  year   month
0        Имя  Количество человек   Год   Месяц
1  Александр                 253  2015  январь
2     Михаил                 247  2015  январь
3       Иван                 226  2015  январь
4     Максим                 208  2015  январь

Удаляем и первую строчку тоже:

In [35]:
df2 = df2.drop(index=[0])
df2.head(5)

name numberofpersons  year   month
1  Александр             253  2015  январь
2     Михаил             247  2015  январь
3       Иван             226  2015  январь
4     Максим             208  2015  январь
5      Артём             190  2015  январь

Проверим наличие пропусков в каждой из колонок:

In [36]:
df2.isna().mean()

name               0.0
numberofpersons    0.0
year               0.0
month              0.0
dtype: float64

Пропусков нет.

Проверим неявные дубликаты в колонке "year":

In [37]:
df2['year'].value_counts()


2015    1200
2016    1200
2017    1200
2018    1200
2019    1200
2020    1200
2021    1200
2022     700
Name: year, dtype: int64

Неявных дубликатов нет. Аналогичным образом проверим колонку "month":

In [38]:
df2['month'].value_counts()

август      500
май         500
июнь        500
июль        500
январь      400
Апрель      400
Март        400
Февраль     400
Январь      400
Декабрь     400
Ноябрь      400
Октябрь     400
апрель      400
март        400
февраль     400
сентябрь    400
декабрь     300
ноябрь      300
октябрь     300
Май         300
Июнь        300
Июль        300
Сентябрь    300
Август      200
Name: month, dtype: int64

Неявные дубликаты есть. Заменим все значения аналогично первой базе данных на названия месяцев русскими прописными буквами

In [39]:
df2.loc[:, 'month'] = df2.loc[:, 'month'].str.lower()
df2['month'].value_counts()

январь      800
февраль     800
март        800
апрель      800
май         800
июнь        800
июль        800
август      700
сентябрь    700
октябрь     700
ноябрь      700
декабрь     700
Name: month, dtype: int64

С колонкой "name" такие же сложности, ка ки с данными по женским именам.

Проверим на наличие явных дубликатов:




In [40]:
df2.duplicated().sum()

0

Явных дубликатов нет. И опять, перед тем как приступить к выполнению задач, надо не забыть поменять типы данных для столбцов "year", "numberofpersons"

In [41]:
df2["year"] = df2["year"].astype(int)
df2["numberofpersons"] = df2["numberofpersons"].astype(int)

## Проверить истинность утверждения «Самыми популярнымимужским детскими именами на протяжении последних трех лет является Александр».
Тут код ничем не будет отличаться от того, который мы писали для проверки схожего утверждения про имя "София/Софья". Только на всякий случай проверим нет ли сокращенной формы "Саша":

In [42]:
df2.query('name	 == "Саша"').name.count()

0

Такая форма имени не встречается. Пишем остальной код:

In [43]:
a = 2022
b = 2021
c = 2020
df2_new2 = df2.query('year == @a or year == @b or year == @c')
df2_new3 = pd.pivot_table(df2_new2,
                values=["numberofpersons"],
                index =["name"],
                aggfunc=[np.sum])
df2_new4 = df2_new3.sort_values([('sum', 'numberofpersons')])
df2_new4.tail(1)


sum
          numberofpersons
name                     
Александр            6432

Имя Александр, если рассматривать сразу 3 года действительно самое популярное. Но как и в случае с женскими именами, нельзя гарантировать, что никакое другое имя не имеет неявных дубликатов в виде сокращенных форм, и если просуммировать эти значения, оно не превзойдет количетсво детей с именем "Александр".
<br>
<br>
Также как и с первой статистикой, посчитаем за каждый год отдельно. Код будет аналогичным.

In [44]:
a = 2022
b = 2021
c = 2020
def count_name(a):
    df2_new2 = df2.query('year == @a')
    df2_new3 = pd.pivot_table(df2_new2,
                values=["numberofpersons"],
                index =["name"],
                aggfunc=[np.sum])
    df2_new4 = df2_new3.sort_values([('sum', 'numberofpersons')])
    return df2_new4.tail(1)

print(count_name(a))
print(count_name(b))
print(count_name(c))

                      sum
          numberofpersons
name                     
Александр            1379
                      sum
          numberofpersons
name                     
Александр            2532
                      sum
          numberofpersons
name                     
Александр            2521


В каждый из трех последних лет имя Александр действительно заняло первое место по популярности. ТАким образом можно сказать, что утверждение правдиво.
<br>
<br>
## Вычислить на сколько процентов изменилась популярность мужского имени Сергей  с начала 2022 года по сравнению с аналогичным периодом 2021 года? 
<br>
Эта задача ничуть не отличается от той, которую мы решали с именем "Есения". Так что и код не будет отличаться:

In [45]:
a = 2022
df2_new1 = df2.query('year == @a and name == "Сергей"')
ans = df2_new1["numberofpersons"].sum()
print(ans)

372


Итого 372 ребенка с именем "Сергей" за это период. посчитаем за 2021 год:

In [46]:
a = 2021
df2_new1 = df2.query('year == @a and name == "Сергей" and month != "август" and month != "сентябрь" and month != "октябрь" and month != "ноябрь" and month != "декабрь"')
ans = df2_new1["numberofpersons"].sum()
print(ans)

337


За период 2021 года с января по июль 337 детей с именем "Сергей".

337 = 100%
372 = x%

x = 372 * 100 / 337

In [47]:
print(372 * 100 / 337)

110.38575667655786


Популярность имени "Сергей" возрасла на 10.4 процента.
<br>
<br>
Мне не совсем понятна формулировка поставленной задачи, поэтому я посчитаю еще суммарно (т.е. сложу количество имен "Сергей" и "Есеня" за 2021 год с января по июль, приму это за 100 процентов и сравню с аналогичной суммой за 2022 год с января по июль)

In [48]:
print(337 + 309)
print(372 + 361)

646
733


646 = 100%
733 = x%

x = 733 * 100 / 646

In [49]:
print(733 * 100 / 646)

113.46749226006192


Итого, если смотерть суммарно, то процент возрос на 13.47
<br>
<br>
## Вывод
Итак, моя работа разбилась на 4 глобальных части: предобработка данных по женским именам, анализ данных по женским именам, предобработка статистики по мужским именам и аналих этих данных.
<br>
<br>
Предобработка данных по мужским и женским именам не сильно отличалась. Я избавилась от ненужных и пустых столбцов и строки. Также я удалила неявные дубликаты, проверила на явные. Помимо этого я изменила тип данных нескольких столбцов для дальнейшего анализа.
<br>
<br>
Сам анализ был также разделен на задачи. В основном я испольщовала фильтры для их решения. С их помощью я выбирала нужные значения из всех имеющихся, а потом проводила с ними дальнейшии действия. Таким образом мне удалось найти наиболее популярные имена, посчитать количество людей названных определенным именем за определенный период и другое. 
